File: ruby_version/lib/customer.rb
class Customer
  attr_reader :type, :arrival_time, :items, :served

  def initialize(hash)
    @type   = hash[:type]
    @items  = hash[:items]
    @arrival_time = hash[:arrival_time]
    @served = false
  end

  def get_in_line(store)
    case @type
    when TYPE_A
      register = store.shortest_line
    when TYPE_B
      register = store.last_customer_with_least_items
    end
    register.line << self
  end

  def ring_item(speed)
    @items -= speed
    @served = true if @items <= 0
  end

end

File: java_version/grocery_stores/src/Customer.java

public class Customer implements Comparable<Customer>{
    public String type;
    public double items;
    public int arrivalTime;
    public boolean served;

    public Customer(String type, int arrivalTime, double items) {
      this.type   = type;
      this.items  = items;
      this.arrivalTime = arrivalTime;
      served = false;
    }

    public void ringItem(double speed) {
      items -= speed;
      if (items <= 0) {
        served = true;
      }
    }

    public void getInLine(Store store) {
        if (type.equals("A")) {
            Register registerA = (Register) store.shortestLine();
            registerA.addCustomer(this);

        } else if (type.equals("B")) {
            Register registerB = (Register) store.lastCustomerWithLeastItems();
            registerB.addCustomer(this);
        }
    }

    @Override
    public int compareTo(Customer customer) {
        int value = Double.compare(items, customer.items);
        return (value == 0) ? type.compareTo(customer.type) : value;
    }
}
